plugin/src/fv_converter/image_feature.cpp

```diff
@@ -0,0 +1,162 @@
+// Jubatus: Online machine learning framework for distributed environment
+// Copyright (C) 2011 Preferred Networks and Nippon Telegraph and Telephone Corporation.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License version 2.1 as published by the Free Software Foundation.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+
+
+#include "image_feature.hpp"
+
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "jubatus/util/data/string/utility.h"
+#include "jubatus/core/fv_converter/exception.hpp"
+#include "jubatus/core/fv_converter/util.hpp"
+
+namespace jubatus {
+namespace plugin {
+namespace fv_converter {
+
+image_feature::image_feature(
+  const std::string& algorithm,
+  const bool resize,
+  float x_size,
+  float y_size)
+  : algorithm_(algorithm),
+  resize_(resize),
+  x_size_(x_size),
+  y_size_(y_size) {
+  if (x_size_ <= 0 || y_size_<= 0) {
+    throw JUBATUS_EXCEPTION(
+        converter_exception("image size must be a positive number"));
+  }
+}
+
+void image_feature::dense_sampler(
+  const cv::Mat mat,
+  const int step,
+  std::vector<cv::KeyPoint>& kp_vec) const {
+  for (int y = step; y < mat.rows-step; y += step) {
+    for (int x = step; x < mat.cols-step; x += step) {
+      kp_vec.push_back(cv::KeyPoint(static_cast<float>(x),
+                                    static_cast<float>(y),
+                                    static_cast<float>(step)));
+    }
+  }
+}
+
+
+void image_feature::add_feature(
+  const std::string& key,
+  const std::string& value,
+  std::vector<std::pair<std::string, float> >& ret_fv) const {
+  std::vector<unsigned char> buf(value.begin(), value.end());
+
+#if(CV_MAJOR_VERSION == 3)
```

Comment:
If ``CV_MAJOR_VERSION`` is not defined, the plugin should be compiled.  For example:

```
#ifndef CV_MAJOR_VERSION
#error "CV_MAJOR_VERSION is undefined"
#endif
```